paasta_tools/autoscaling/autoscaling_cluster_lib.py

```diff
@@ -364,69 +364,97 @@ def scale_aws_spot_fleet_request(resource, current_capacity, target_capacity, po
+            raise
+    log.info("Decreasing spot fleet capacity from {0} to: {1}".format(current_capacity, new_capacity))
+    try:
+        set_spot_fleet_request_capacity(sfr_id, new_capacity, dry_run, region=resource['region'])
+    except FailSetSpotCapacity:
+        log.error("Couldn't update spot fleet, stopping autoscaler")
+        log.info("Undraining {0}".format(slave_to_kill['pid']))
+        if not dry_run:
+            undrain([drain_host_string])
+        raise
+    log.info("Waiting for instance to drain before we terminate")
+    try:
+        wait_and_terminate(slave_to_kill, drain_timeout, dry_run, region=resource['region'])
+    except ClientError as e:
+        log.error("Failure when terminating: {0}: {1}".format(slave_to_kill['pid'], e))
+        log.error("Setting spot fleet capacity back to {0}".format(current_capacity))
+        set_spot_fleet_request_capacity(sfr_id, current_capacity, dry_run, region=resource['region'])
+    finally:
+        log.info("Undraining {0}".format(slave_to_kill['pid']))
+        if not dry_run:
+            undrain([drain_host_string])
+
+
+def downscale_spot_fleet_request(resource, filtered_slaves, current_capacity, target_capacity, pool_settings, dry_run):
+    while True:
+        filtered_sorted_slaves = sort_slaves_to_kill(filtered_slaves)
+        if len(filtered_sorted_slaves) == 0:
+            break
+        log.info("SFR slave kill preference: {0}".format([slave['hostname'] for slave in filtered_sorted_slaves]))
+        filtered_sorted_slaves.reverse()
+        slave_to_kill = filtered_sorted_slaves.pop()
+        # Instance weights can be floats but the target has to be an integer
+        # because AWS...
+        instance_capacity = slave_to_kill['instance_weight']
+        new_capacity = int(round(current_capacity - instance_capacity))
+        if new_capacity < target_capacity:
+            log.info("Terminating instance {0} with weight {1} would take us below our target of {2}, so this is as"
+                     " close to our target as we can get".format(slave_to_kill['instance_id'],
+                                                                 slave_to_kill['instance_weight'],
+                                                                 target_capacity))
+            break
+        try:
+            gracefully_terminate_slave(resource=resource,
+                                       slave_to_kill=slave_to_kill,
+                                       pool_settings=pool_settings,
+                                       current_capacity=current_capacity,
+                                       new_capacity=new_capacity,
+                                       dry_run=dry_run)
+        except HTTPError:
+            # Something wrong draining host so try next host
+            continue
+        except FailSetSpotCapacity:
+            break
+        current_capacity = new_capacity
```

Comment:
I suppose it could be set within the `gracefully_terminate_slave` function but it seemed easier to be outside and I put it at the end to make sure it only got updated if we definitely succeeded in changing the capacity.